### ISSUE TITLE
sophia: duplicate-key error followed by segmentation fault

### ISSUE THREAD

**Issue by pgulutzan** [tarantool/tarantool#698]
I have Ubuntu 12.04 and Tarantool 1.6, pulled from source today.

I start Tarantool 1.6 on  an empty directory, that is, I say 'rm -R *' before I start.

I create a function which will insert or delete or replace or update random
20-byte strings in a single space, 100000 times.
I call this function 5 times with different random seed values.
If the space is defined with engine_name='memtx', no problem.
If the space is defined with engine_name='sophia', the display includes lines like these:

```
tarantool> math.randomseed(1)
---
...

tarantool> delete_replace_update()
2015-01-25 11:35:25.499 [26032] wal I> creating `./00000000000000000000.xlog.inprogress'
---
- error: Duplicate key exists in unique index 0
...

tarantool> print (counter, random_number, string_value_2, string_value_3)
83113	3	IBROOHLHAUMIFZWRFDIN	RNEANASZYWLSZDFVHAFE
---
...

tarantool> 
---
...

tarantool> math.randomseed(2)
---
...

tarantool> delete_replace_update()
Segmentation fault
Current time: 1422210934
Please file a bug at http://github.com/tarantool/tarantool/issues
Attempting backtrace... Note: since the server has already crashed, 
this may fail as well
2015-01-25 11:35:34.713 [26034] main/101/spawner I> Exiting: master shutdown
Segmentation fault (core dumped)
```

The duplicate-key error is happening in code that does delete followed by insert,
during iteration #83113. The chances of getting a real duplicate in such a short time,
for such a long key, seem absurdly low to me. Besides, as I said, it doesn't happen
with 'memtx'.

How to repeat:

The following script will create a directory named tarantool_sandbox,
erase everything on it, start Tarantool, create the random-string-generator
and insert-replace-delete-update function, call the latter function 5 times
with different seeds. After each call it displays the number of iterations
(which will be less than 100000 if there is a duplicate-key error), the type
of operation, and the string values being used during the last operation.
It will fail. To run the test again with the 'memtx' engine, change the line
engine_name = 'sophia'
to
engine_name = 'memtx'
I apologize that due to time constraint I did not make a nicer script.

```
mkdir ~/tarantool_sandbox
cd ~/tarantool_sandbox
rm -R *
~/tarantool-master/src/tarantool
box.cfg{}
engine_name = 'sophia'
console = require('console'); console.delimiter('!')
function string_function()
  local random_number
  local random_string
  random_string = ""
  for x = 1,20,1 do
    random_number = math.random(65, 90)
    random_string = random_string .. string.char(random_number)
    end
  return random_string
  end!
function delete_replace_update()
  local string_value
  if (box.space._space.index.name:select{'tester'}[1] ~= nil) then
    box.space.tester:drop()
    end
  box.schema.space.create('tester', {engine=engine_name})
  box.space.tester:create_index('primary',{type = 'tree', parts = {2, 'STR'}})
  counter = 1
  while counter < 100000 do
    string_value = string_function()
    string_table = box.space.tester.index.primary:select({string_value}, {iterator = 'GE', limit = 1})
    if string_table[1] == nil then string_value_2 = box.space.tester:insert{counter, string_value}[2]
    else string_value_2 = string_table[1][2] end
    if string_value_2 == nil then string_value_2 = box.space.tester:insert{counter, string_value}[2] end
    random_number = math.random(1,6)
    string_value_3 = string_function()
    if random_number == 1 then
      box.space.tester:delete{string_value_2}
      end
    if random_number == 2 then
      box.space.tester:replace{counter, string_value_2, string_value_3}
      end
    if random_number == 3 then
      box.space.tester:delete{string_value_2}
      box.space.tester:insert{counter, string_value_2}
      end
    if random_number == 4 then
      if box.space.tester:len() < 1000000 then
        box.space.tester:delete{string_value_3}
        box.space.tester:insert{counter, string_value_3, string_value_2}
        end
      end
    if random_number == 5 then
      box.space.tester:update({string_value_2}, {{'=', 1, string_value_3}})
      end
    if random_number == 6 then
      box.space.tester:update({string_value_2}, {{'=', 2, string_value_3}})
      end
    counter = counter + 1
    end
  print('OK')
  end!
console.delimiter('')!
math.randomseed(1)
delete_replace_update()
print (counter, random_number, string_value_2, string_value_3)
ath.randomseed(2)
delete_replace_update()
print (counter, random_number, string_value_2, string_value_3)
math.randomseed(3)
delete_replace_update()
print (counter, random_number, string_value_2, string_value_3)
math.randomseed(4)
delete_replace_update()
print (counter, random_number, string_value_2, string_value_3)
math.randomseed(5)
delete_replace_update()
print (counter, random_number, string_value_2, string_value_3)
```

**Comment by pmwkaa** [comment]
I've tried to run this test script a couple of times but could not reproduce:

```
tarantool> dofile('test.lua')
2015-01-26 15:27:57.741 [10885] wal I> creating `./00000000000000583502.xlog.inprogress'
OK
100000  3       EISYZSTXKHBIALYHRHCE    BUPYVKWEHVEFKSZFAQIK
OK
100000  3       DTZNAFACLDNDJDZSJYBM    UHMBETMEPMQWOIHLAOSR
OK
100000  1       KORRWYSPZQSWVGKQQMUN    RSMFJYXLNSOILKDEFJZW
OK
100000  3       OKLKZHUBWTMIJIXZSNPR    RXXYWNPDXVVGONOGTEZK
2015-01-26 15:28:35.123 [10885] wal I> creating `./00000000000001083502.xlog.inprogress'
OK
100000  3       LCWTFMIHEVZCCRBYAYZI    HKBELZCRRWNZHFIRAHJD
---
...

tarantool> dofile('test.lua')
OK
100000  3       EISYZSTXKHBIALYHRHCE    BUPYVKWEHVEFKSZFAQIK
OK
100000  3       DTZNAFACLDNDJDZSJYBM    UHMBETMEPMQWOIHLAOSR
OK
100000  1       KORRWYSPZQSWVGKQQMUN    RSMFJYXLNSOILKDEFJZW
2015-01-26 15:34:21.459 [10885] wal I> creating `./00000000000001583502.xlog.inprogress'
OK
100000  3       OKLKZHUBWTMIJIXZSNPR    RXXYWNPDXVVGONOGTEZK
OK
100000  3       LCWTFMIHEVZCCRBYAYZI    HKBELZCRRWNZHFIRAHJD
---
...

tarantool> dofile('test.lua')
OK
100000  3       EISYZSTXKHBIALYHRHCE    BUPYVKWEHVEFKSZFAQIK
OK
100000  3       DTZNAFACLDNDJDZSJYBM    UHMBETMEPMQWOIHLAOSR
2015-01-26 15:35:44.549 [10885] wal I> creating `./00000000000002083502.xlog.inprogress'
OK
100000  1       KORRWYSPZQSWVGKQQMUN    RSMFJYXLNSOILKDEFJZW
OK
100000  3       OKLKZHUBWTMIJIXZSNPR    RXXYWNPDXVVGONOGTEZK
OK
100000  3       LCWTFMIHEVZCCRBYAYZI    HKBELZCRRWNZHFIRAHJD
---
...

tarantool> 
```

But i've found that test script tries to successfully update a primary key. This operation should be banned by tarantool, and i believe could be the cause of the issue.

I've posted a ticket: gh-699

```
 if random_number == 6 then
      box.space.tester:update({string_value_2}, {{'=', 2, string_value_3}})
 end
```

**Comment by pgulutzan** [comment]
It's up to you to ban changes of primary key, but I removed those lines, ran again, and still saw a duplicate-key error followed by a segmentation fault. Is it possible that I am not using the correct sophia version (I said "git submodule update" when I pulled but perhaps that's not enough)? Or is it possible that our random number generation methods differ in some significant way?

**Comment by kostja** [comment]
I can repeat this.

**Comment by pmwkaa** [comment]
Fixed by the referenced commit

**Comment by pgulutzan** [comment]
Today, I do not see the crash, but I do see the duplicate-key error. The following script will produce a duplicate-key error on my machine, reliably after about 58000 iterations. For example immediately after key USPTCRITLOCBZSEZHJMR is successfully deleted, an attempt to insert with key USPTCRITLOCBZSEZHJMR fails.


mkdir ~/tarantool_sandbox
cd ~/tarantool_sandbox
rm -R *
~/tarantool-master/src/tarantool
box.cfg{}
engine_name = 'sophia'
console = require('console'); console.delimiter('!')
function string_function()
  local random_number
  local random_string
  random_string = ""
  for x = 1,20,1 do
    random_number = math.random(65, 90)
    random_string = random_string .. string.char(random_number)
    end
  return random_string
  end!
function delete_replace_update()
  local string_value
  if (box.space._space.index.name:select{'tester'}[1] ~= nil) then
    box.space.tester:drop()
    end
  box.schema.space.create('tester', {engine=engine_name})
  box.space.tester:create_index('primary',{type = 'tree', parts = {2, 'STR'}})
  counter = 1
  while counter < 1000000 do
    string_value = string_function()
    string_table = box.space.tester.index.primary:select({string_value}, {iterator = 'GE', limit = 1})
    if string_table[1] == nil then
print (1, ' insert', counter, string_value)
      string_value_2 = box.space.tester:insert{counter, string_value}[2]
      else
      string_value_2 = string_table[1][2]
      end
    if string_value_2 == nil then
print (2, ' insert', counter, string_value)
      string_value_2 = box.space.tester:insert{counter, string_value}[2]
      end

print (3, ' delete', counter, string_value_2)
      box.space.tester:delete{string_value_2}
print (4, ' insert', counter, string_value_2)
      box.space.tester:insert{counter, string_value_2}
    counter = counter + 1
    end
  print('OK')
  end!
console.delimiter('')!

delete_replace_update()

**Comment by pmwkaa** [comment]
Couldn't repeat, please check that you are using 'sophia-integration' branch and sophia submodule points to the latest commit (i think that the problem).

I've added test case with some fixes: sophia index doesn't return old tuple on replace (due to performance consideration). Thank you

**Comment by kostja** [comment]
The test still fails.

**Comment by kostja** [comment]
Fixed in 1.6.4-637-g5ef2b15
